src/EPR.Calculator.API/Builder/Summary/SaSetupCosts/SaSetupCostsProducer.cs

```diff
@@ -0,0 +1,240 @@
+                calcResult.CalcResultParameterOtherCost.BadDebtValue / 100;
+        }
+
+        public static decimal GetProducerOneOffFeeWithBadDebtProvision(IEnumerable<ProducerDetail> producers,
+            IEnumerable<MaterialDetail> materials,
+            CalcResult calcResult,
+            Dictionary<MaterialDetail, CalcResultSummaryProducerDisposalFeesByMaterial> materialCostSummary,
+            Dictionary<MaterialDetail, CalcResultSummaryProducerCommsFeesCostByMaterial> materialCommsCostSummary)
+        {
+            return GetProducerOneOffFeeWithoutBadDebtProvision(producers, materials, calcResult, materialCostSummary, materialCommsCostSummary) +
+                GetBadDebtProvision(producers, materials, calcResult, materialCostSummary, materialCommsCostSummary);
+        }
+
+        public static decimal GetProducerOneOffFeeWithoutBadDebtProvisionTotal(IEnumerable<ProducerDetail> producers,
+            IEnumerable<ProducerDetail> producersAndSubsidiaries,
+            IEnumerable<MaterialDetail> materials,
+            CalcResult calcResult)
+        {
+            var oneOffFeeSetupCostsWithoutBadDebtProvision = SaSetupCostsSummary.GetOneOffFeeSetupCostsWithoutBadDebtProvision(calcResult);
+
+            var total1Plus2ABadDebtForAllProducers = CalcResultSummaryUtil.GetTotal1Plus2ABadDebt(producers, materials, calcResult);
+
+            var total1Plus2ABadDebtForProducersAndSubsidiaries = CalcResultSummaryUtil.GetTotal1Plus2ABadDebt(producersAndSubsidiaries, materials, calcResult);
+
+            var producerPercentageOfOverallProducerCosts = total1Plus2ABadDebtForProducersAndSubsidiaries / total1Plus2ABadDebtForAllProducers;
+
+            return producerPercentageOfOverallProducerCosts * oneOffFeeSetupCostsWithoutBadDebtProvision;
+        }
+
+        public static decimal GetBadDebtProvisionTotal(IEnumerable<ProducerDetail> producers,
+            IEnumerable<ProducerDetail> producersAndSubsidiaries,
+            IEnumerable<MaterialDetail> materials,
+            CalcResult calcResult)
+        {
+            return GetProducerOneOffFeeWithoutBadDebtProvisionTotal(producers, producersAndSubsidiaries, materials, calcResult) *
+                calcResult.CalcResultParameterOtherCost.BadDebtValue / 100;
+        }
+
+        public static decimal GetProducerOneOffFeeWithBadDebtProvisionTotal(IEnumerable<ProducerDetail> producers,
+            IEnumerable<ProducerDetail> producersAndSubsidiaries,
+            IEnumerable<MaterialDetail> materials,
+            CalcResult calcResult)
+        {
+            return GetProducerOneOffFeeWithoutBadDebtProvisionTotal(producers, producersAndSubsidiaries, materials, calcResult) +
+                GetBadDebtProvisionTotal(producers, producersAndSubsidiaries, materials, calcResult);
+        }
+
+        public static decimal GetEnglandTotalWithBadDebtProvision(IEnumerable<ProducerDetail> producers,
```

Comment:
There's a lot of duplicated code here - could it be refactored to reduce duplication?

Reply:
Thank you @tim-powell-eviden, as discussed I will leave it for now as the functions contain one-liner return statements and we will further improve the code later if required.